django/test/testcases.py

```diff
@@ -776,6 +777,21 @@ def assertFieldOutput(self, fieldclass, valid, invalid, field_args=None,
             field_kwargs.update({'min_length': 2, 'max_length': 20})
             self.assertIsInstance(fieldclass(*field_args, **field_kwargs), fieldclass)
 
+    def assertQueryEqual(self, query1, query2, msg=None):
+        """
+        Assert that two queries generate same sql. .
+        The arguments must be valid Query.
```

Comment:
Sorry I didn't provide a better tip -- I was talking about Django's own testing tool for this: https://docs.djangoproject.com/en/dev/topics/testing/tools/#django.test.TransactionTestCase.assertQuerysetEqual

Have a look at the other test cases that use this test method.

Reply:
@jacobtylerwalls okay! Thanks!

Reply:
@jacobtylerwalls 

I wanna make two testcases.

1) 
        self.assertQueryEqual(
            Item.objects.exclude(modified__year='2021').values('id').query,
            Item.objects.annotate(year=ExtractYear('modified')).exclude(year='2021').values('id').query,
        )
it is enough to write code like below, right?
self.assertEqual(
    str(Item.objects.exclude(modified__year='2021').values('id').query),
    str(tem.objects.annotate(year=ExtractYear('modified')).exclude(year='2021').values('id').query)
)
for proving case about the reported issue.

2) check the IS NOT NULL includes or not in about each cases.
about this cases  I will use 'TransactionTestCase.assertQuerysetEqual' this.

Is it okay?

Reply:
Go for it. Another reviewer may let you know if it looks unnecessary to test the raw SQL, but we do have other tests that do this: https://github.com/django/django/blob/a948d9df394aafded78d72b1daa785a0abfeab48/tests/queries/tests.py#L3137-L3139

I was just trying to encourage you to create and evaluate querysets. 👍🏻

Reply:
'TransactionTestCase.assertQuerysetEqual' compare query result(return data from db) with iterable. that is not test i want. because i wanna check to include  'IS NOT NULL' or not according to each condition.  give me your opinion :)

Reply:
I'm just one voice -- if other reviewers think testing the raw SQL is fine, we can roll on. Sure, let's get your test of the raw SQL in place and then hear from other reviewers.

What makes me nervous about only verifying the raw SQL instead of checking that the query got the right objects from the db is that these queries are complex -- how do I know the IS NOT NULL is in the right place? Won't there be a false negative if there's an IS NOT NULL on some other column? Then to review your patch I have to print out the raw SQL and check it myself, if you get where I'm going.

Reply:
@jacobtylerwalls What do you think about checking entire SQL more explicitly?

Reply:
I added only case1. I think it's enough to prove the issue.  if you or other reviewers think we need more tests, tell me.